pkg/kubelet/kubelet.go

```diff
@@ -929,34 +929,107 @@ type Kubelet struct {
+	// * terminated: pod should have all resources cleaned up (syncTerminatedPod)
+	//
+	// and invoke the handler methods that correspond to each state. Components within the
+	// kubelet that need to know the phase of the pod in order to correctly set up or tear down
+	// resources must consult the podWorkers.
+	//
+	// Once a pod has been accepted by the pod workers, no other pod with that same UID (and
+	// fullName, for static pods) will be started until the first pod has fully terminated and
+	// been cleaned up by SyncKnownPods. This means a pod may be desired (in API), admitted (in
+	// pod manager), and requested (by invoking UpdatePod) but not start for an arbitrarily long
+	// interval because a prior pod is still terminating.
+	//
+	// As an event-driven (by UpdatePod) controller, the podWorkers must periodically be resynced
+	// by the kubelet invoking SyncKnownPods with the desired state (admitted pods in podManager).
+	// Since the podManager may be unaware of some running pods due to force deletion, the
+	// podWorkers are responsible for triggering a sync of pods that are no longer desired but
+	// must still run to completion.
 	podWorkers PodWorkers
 
+	// evictionManager observes the state of the node for situations that could impact node stability
+	// and evicts pods (sets to phase Failed with reason Evicted) to reduce resource pressure. The
+	// eviction manager acts on the actual state of the node and considers the podWorker to be
+	// authoritative.
+	evictionManager eviction.Manager
+
+	// probeManager tracks the set of running pods and ensures any user-defined periodic checks are
+	// run to introspect the state of each pod.  The probe manager acts on the actual state of the node
+	// and is notified of pods by the podWorker. The probe manager is the authorative source of the
+	// most recent probe status and is responsible for notifying the status manager, which
+	// synthesizes them into the overall pod status.
+	probeManager prober.Manager
+
+	// secretManager caches the set of secrets used by running pods on this node. The podWorkers
+	// notify the secretManager when pods are started and terminated, and the secretManager must
+	// then keep the needed secrets up-to-date as they change.
+	secretManager secret.Manager
+
+	// configMapManager caches the set of config maps used by running pods on this node. The
+	// podWorkers notify the configMapManager when pods are started and terminated, and the
+	// configMapManager mustthen keep the needed config maps up-to-date as they change.
+	configMapManager configmap.Manager
+
+	// volumeManager observes the set of running pods and is responsible for attaching, mounting,
+	// unmounting, and detaching as those pods move through their lifecycle. It periodically
+	// synchronizes the set of known volumes to the set of actually desired volumes and cleans up
+	// any orphaned volumes. The volume manager considers the podWorker to be authoritative for
+	// which pods are running.
```

Comment:
re: 

> The volume manager considers the podWorker to be authoritative for which pods are running

looking at volume manager `desiredStateOfWorldPopulator`, it seems like it is using pod manager:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager.go#L269-L271

and xref:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L205-L220

But within `desiredStateOfWorldPopulator` it looks like it consults pod worker with `ShouldPodContainersBeTerminating` using `podStateProvider` interface which uses `podWorkers` underneath: https://github.com/kubernetes/kubernetes/blob/703361f381d4a1eaa7dc87ff3b7c12c84f3878a4/pkg/kubelet/kubelet.go#L819

So tl;dr to me it looks like volumeManager is kind of using a hybrid between pod manager and pod worker, and it sounds like another opportunity in future to unify and just use pod worker as a single authoritative source.

Reply:
Right - I caught the problem (the first time around) of the volume manager trying to guess what pods are running (and redirected it to the source of truth, the worker).  But I completely missed that the desired set of pods for the volume manager is running pods, not an upstream of running pods.